src/service/s3.rs

```diff
@@ -0,0 +1,54 @@
+use super::name::{to_block, to_block_raw};
+use anyhow::Result;
+use libipld::{cid::Cid, Block, DagCbor, DefaultParams};
+use std::collections::BTreeMap;
+
+#[derive(DagCbor, PartialEq, Debug, Clone)]
+pub struct S3Object {
+    pub key: Vec<u8>,
+    pub value: Cid,
+    pub metadata: BTreeMap<String, String>,
+}
+
+impl S3Object {
+    pub fn new(
+        key: Vec<u8>,
+        value: Cid,
+        metadata: impl IntoIterator<Item = (String, String)>,
+    ) -> Self {
+        Self {
+            key,
+            value,
+            metadata: metadata.into_iter().collect(),
+        }
+    }
+
+    pub fn to_block(&self) -> Result<Block<DefaultParams>> {
+        to_block(self)
+    }
+}
+
+pub struct S3ObjectBuilder {
+    pub key: Vec<u8>,
+    pub metadata: BTreeMap<String, String>,
+}
+
+impl S3ObjectBuilder {
+    pub fn new(key: Vec<u8>, metadata: impl IntoIterator<Item = (String, String)>) -> Self {
```

Comment:
I don't think the Iterator part is needed, looking at the current usage it can simply accept a `Vec`, and let the caller `clone` it if needed. Applies to the other functions.